### ISSUE TITLE
Error with latest Docker build

### ISSUE THREAD

**Issue by brettChapman** [pangenome/pggb#42]
Hi Erik

I just updated to the latest Docker build of PGGB: 20201210144337a5a404

As soon as Edyeet starts to run I get this error:
```
+ srun -n 1 singularity exec --bind /data/pangenome_20way/edyeet_results/2H:/data/pangenome_20way/edyeet_results/2H /data/pggb_builds/pggb.sif pggb -i /data/pangenome_20way/edyeet_results/2H/all_genomes_1640.tmp -w 10000 -s 1000000 -I 0 -p 85 -a 85 -n 18 -t 16 -v -l -S -o barley_pangenome_2H_s1000000_p85_a85_split
Command terminated by signal 4
edyeet -X -s 1000000 -p 85 -n 18 -a 85 -k 16 -t 16 /data/pangenome_20way/edyeet_results/2H/all_genomes_1640.tmp /data/pangenome_20way/edyeet_results/2H/all_genomes_1640.tmp
0.00s user 0.03s system 6% cpu 0.55s total 5424Kb max memory
srun: error: node-7: task 0: Exited with exit code 132
```

Edyeet does't seem to be detecting the correct memory available. This only happened immediately after I updated to the latest PGGB build. I'll see if rolling back to the version a week ago will help for now. Thanks.

**Comment by brettChapman** [comment]
I'm just trying with build: 20201130143807d2fc2a (which is from 11 days ago)

Seems to be running ok so far using a variety of parameters I've used before with success: a80/p80, a85/p85, a90/p90 segment lengths of 100Kbp, 150Kbp, 1Mbp, and starts running -N (but will likely die later with -N, not immediately as with your latest build). Before when the jobs died, it was usually quite far along into the alignment or at the smoothxg step with a max memory error of 130GB (I have 128GB RAM now), not immediately as Edyeet starts as it was doing here with your latest build.

**Comment by brettChapman** [comment]
Also worth noting, I tried with a build from 9 days ago (20201201231633a17703, 202012021223060cd3eb), and the jobs immediately died with the same error.

**Comment by brettChapman** [comment]
Hi Erik

The behavior of PGGB seems to have changed a lot. After running with PGGB build 20201130143807d2fc2a over the weekend, the alignment fails at Edyeet and then seqwish. Using these exact parameters I had no problems getting through the entire workflow before. I'll try rolling back the build further still and see how it performs. Using other more loose parameters (lower -a/-p values) I also had the same problem. 

```
srun -n 1 singularity exec --bind /data/pangenome_20way/edyeet_results/2H:/data/pangenome_20way/edyeet_results/2H /data/pggb_builds/pggb.sif pggb -i /data/pangenome_20way/edyeet_results/2H/all_genomes_1648.tmp -w 10000 -s 150000 -I 0 -p 90 -a 90 -n 18 -t 16 -v -l -S -o barley_pangenome_2H_s150000_p90_a90_split
[edyeet::map] Reference = [/data/pangenome_20way/edyeet_results/2H/all_genomes_1648.tmp]
[edyeet::map] Query = [/data/pangenome_20way/edyeet_results/2H/all_genomes_1648.tmp]
[edyeet::map] Kmer size = 16
[edyeet::map] Window size = 6000
[edyeet::map] Segment length = 150000 (read split allowed)
[edyeet::map] Alphabet = DNA
[edyeet::map] Percentage identity threshold = 90%
[edyeet::map] Mapping output file = /data/pangenome_20way/edyeet_results/2H/edyeet-Y3SyWa
[edyeet::map] Filter mode = 1 (1 = map, 2 = one-to-one, 3 = none)
[edyeet::map] Execution threads  = 16
[edyeet::skch::Sketch::build] minimizers picked from reference = 4667424
[edyeet::skch::Sketch::index] unique minimizers = 196457
[edyeet::skch::Sketch::computeFreqHist] Frequency histogram of minimizers = (1, 61793) ... (49827, 1)
[edyeet::skch::Sketch::computeFreqHist] With threshold 0.001%, ignore minimizers occurring >= 49827 times during lookup.
[edyeet::map] time spent computing the reference index: 52.7955 sec
[edyeet::skch::Map::mapQuery] mapped 100.00% @ 2.42e+06 bp/s elapsed: 00:01:32:26 remain: 00:00:00:00
[edyeet::skch::Map::mapQuery] count of mapped reads = 20, reads qualified for mapping = 20, total input reads = 20, total input bp = 13440953369
[edyeet::map] time spent mapping the query: 5.55e+03 sec
[edyeet::map] mapping results saved in: /data/pangenome_20way/edyeet_results/2H/edyeet-Y3SyWa
[edyeet::align] Reference = [/data/pangenome_20way/edyeet_results/2H/all_genomes_1648.tmp]
[edyeet::align] Query = [/data/pangenome_20way/edyeet_results/2H/all_genomes_1648.tmp]
[edyeet::align] Mapping file = /data/pangenome_20way/edyeet_results/2H/edyeet-Y3SyWa
[edyeet::align] Edlib identity cut-off = 9.00e+01%
[edyeet::align] Alignment output file = /dev/stdout
[edyeet::align] time spent read the reference sequences: 2.47e+01 sec
[edyeet::align::computeAlignments] aligned 100.00% @ 9.56e+05 bp/s elapsed: 02:05:43:40 remain: 00:00:00:00
[edyeet::align::computeAlignments] count of mapped reads = 20, total aligned bp = 184902750000
[edyeet::align] time spent computing the aligment: 1.93e+05 sec
[edyeet::align] alignment results saved in: /dev/stdout
edyeet -X -s 150000 -p 90 -n 18 -a 90 -k 16 -t 16 /data/pangenome_20way/edyeet_results/2H/all_genomes_1648.tmp /data/pangenome_20way/edyeet_results/2H/all_genomes_1648.tmp
3342166.10s user 259.05s system 1679% cpu 199058.69s total 49820048Kb max memory
[seqwish::seqidx] 0.005 indexing sequences
[seqwish::seqidx] 229.731 index built
[seqwish::alignments] 229.731 processing alignments
[seqwish::alignments] 721.007 indexing
[seqwish::alignments] 2984.010 index built
[seqwish::transclosure] 2984.393 computing transitive closures
[seqwish::transclosure] 2986.423 0.00% 0-1000000 overlap_collect
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Command terminated by signal 6
seqwish -t 16 -s /data/pangenome_20way/edyeet_results/2H/all_genomes_1648.tmp -p barley_pangenome_2H_s150000_p90_a90_split/all_genomes_1648.tmp.pggb-s150000-p90-n18-a90-K16-k8-w10000-j5000-e5000-I0.paf -k 8 -g barley_pangenome_2H_s150000_p90_a90_split/all_genomes_1648.tmp.pggb-s150000-p90-n18-a90-K16-k8-w10000-j5000-e5000-I0.seqwish.gfa -B 1000000 -P
23377.42s user 3592.62s system 268% cpu 10045.38s total 130260836Kb max memory
srun: error: node-5: task 0: Exited with exit code 134
```

**Comment by brettChapman** [comment]
I'm trying out your latest build with -a85/-p85 and -s1000000. So far it's gone further than the initial error I received at the top of this thread:
```
srun -n 1 singularity exec --bind /data/pangenome_20way/edyeet_results/2H:/data/pangenome_20way/edyeet_results/2H /data/pggb_builds/pggb.sif pggb -i /data/pangenome_20way/edyeet_results/2H/all_genomes_1991.tmp -s 1000000 -p 85 -a 85 -n 18 -t 16 -m -v -l -S -o barley_pangenome_2H_s1000000_p85_a85_split
[edyeet::map] Reference = [/data/pangenome_20way/edyeet_results/2H/all_genomes_1991.tmp]
[edyeet::map] Query = [/data/pangenome_20way/edyeet_results/2H/all_genomes_1991.tmp]
[edyeet::map] Kmer size = 16
[edyeet::map] Window size = 10526
[edyeet::map] Segment length = 1000000 (read split allowed)
[edyeet::map] Alphabet = DNA
[edyeet::map] Percentage identity threshold = 85%
[edyeet::map] Mapping output file = /data/pangenome_20way/edyeet_results/2H/edyeet-G4y2zY
[edyeet::map] Filter mode = 1 (1 = map, 2 = one-to-one, 3 = none)
[edyeet::map] Execution threads  = 16
[edyeet::skch::Sketch::build] minimizers picked from reference = 2600670
[edyeet::skch::Sketch::index] unique minimizers = 98549
[edyeet::skch::Sketch::computeFreqHist] Frequency histogram of minimizers = (1, 23864) ... (40963, 1)
[edyeet::skch::Sketch::computeFreqHist] With threshold 0.001%, consider all minimizers during lookup.
[edyeet::map] time spent computing the reference index: 51.061 sec
```

Will see how it goes. Fingers crossed.

**Comment by ekg** [comment]
A big update is coming, specifically in terms of edyeet behavior and
smoothxg performance.

Sorry to keep you waiting. It should all get synced in within the next few
days.

**Comment by brettChapman** [comment]
Thanks Erik. I look forward to trying out the new upcoming update!

**Comment by cistarsa** [comment]
I'm also looking forward to this (experiencing same docker image issue as op).

**Comment by subwaystation** [comment]
Did the latest update fix your issues @brettChapman @cistarsa ?

**Comment by brettChapman** [comment]
@subwaystation Sorry for the late reply. Yes, the update has improved things significantly. I've been running PGGB now for a few weeks, without any issues. I did hit a memory limit with one chromosome though, with virtual memory peaking at 750GB. I plan to increase swap space to get around the problem. I've been running with -a85, -p85, -s1000000. In later runs I'm considering lowering -a to 80 or 75. I see -a is now set to a default of 0. I imagine running with -a0 would increase run time and memory usage significantly though, and would likely include spurious sequences in the alignment, so I'll stick with -a close to the value of -p.

**Comment by subwaystation** [comment]
Thanks for the feedback @brettChapman!

Sounds to me like the issue is solved. If you have a different feeling, please open again.

**Comment by ekg** [comment]
We're tending to use the pggb -l (wfmash/edyeet-related) flag to filter
short mappings.

-a is basically not doing anything for wfmash, and for edyeet it is just
removing things.

The algorithms that are used for mapping have changed a lot from when you
first started testing. We now use a wavefront-based high-level alignment
method in these two. It lets us directly align very long sequences with
pretty significant divergence but without much memory.

We really should set good defaults, but we're still figuring it out.

This presentation describes the settings we used on a recent run for the
HPRC:
https://docs.google.com/presentation/d/1dIuutRY7fSgD88mqAPwP7oTmr1ZAVxHLVnO9iMkqLww/edit?usp=sharing

I'd suggest starting with those settings and seeing how it goes for you.

**Comment by brettChapman** [comment]
Thanks @ekg I've looked through your HPRC PGGB example slides. I'll try implementing the same parameters, but for the segment length -s I think I'll keep that either 100K or 1MBp, as Barley is highly repetitive (~80% repetitive), and I'll keep -l 3x the segment length, and I'll set -a to 0, and set other parameters as in your example. I'll also try running with the entire genome, as I'll be setting my swap space to 2TB now.

I've been running on each chromosome and have reached the smoothxg step for all. I've found in 2 cases however the virtual memory hits around 800GB on chromosome 1 and 7, and the job is killed and will likely be killed for the other jobs. I'm going to rebuild my cluster with 2TB of swap space to finish the jobs off, so I can get some results out for my supervisor and our collaborators. Once I have something working for them in sequenceTubeMap and potentially MoMI-G, I'll revisit building the genome graph again using more appropriate parameters.
